Separate tfbackend configs into separate files

lorenyu opened the pull request:
## Ticket

Resolves #221 
Resolves https://github.com/navapbc/template-infra/issues/258

## Changes
* Separate terraform backend configs into separate *.s3.tfbackend files
* Separate environment configs into [env].tfvar files
* Rename env-template to service to represent application layer (in preparation for adding a db layer)
* Remove envs folders now that the service (previously env-template) module can act directly as the root module
* Add make commands for setting up account, app build repository, and app service
* Add make commands for terraform init/applying the account, app build repository, and app service
* Remove redundant make commands in template-only.mak
* Update template tests
* Update instructions
* Update module architecture documentation
* Update update.sh to not revert main.tf files
* Add ADR documenting the discussion and benefits of this approach
* Update ci-infra

Future todos created
* Add error messages when insufficient parameters passed (will be addressed in https://github.com/navapbc/template-infra/issues/263)

## Context for reviewers
We had a [discussion over 🔒 slack](https://nava.slack.com/archives/C03G1SWD9H7/p1676305850843949) about having terraform backend configs in main.tf only be partial configurations and having the rest of the configuration values (the s3 bucket, the tf state key, and the region) in separate .tfbackend files. This removes the need to have an "env-template" module and separate root modules for each environment. Instead all environments can share a single root module, and infra engineers can decide which environment to work on by passing in a `-backend-config` option to `terraform init` (e.g. `terraform init -backend-config=dev.s3.tfbackend`). A wrapper script would be used for convenience.

In this PR, the top level `Makefile` operates as the wrapper script. Thus to make changes to the dev environment, an engineer would run

```
make infra-app-service ENVIRONMENT=dev
```

the developer can pass additional arguments to terraform apply via the `TF_APPLY_ARGS` parameter e.g.

```
make infra-app-service ENVIRONMENT=dev TF_APPLY_ARGS="-auto-approve -var=image_tag=aaaaaaa"
```

Similarly, if the engineer wants to make changes to the account layer they'd run

```
make infra-account
```

or if they want to make changes to the ECR repo, they'd run

```
make infra-build-repository
```

One of the advantages discussed was to eliminate incidental differences between environments when resources are added to root modules specific to an environment.

Another advantage, and the reason why I started exploring this path, is to avoid needing to add a lot of extra code in order to add the db layer. In particular, when I started adding a db layer, I realized I needed to have a `db-env-template` module, a `db-envs` folder, and a root module for every environment within `db-envs`, and then developing and testing the setup of all those modules would have added quite a bit of extra work.

## Testing
Ran through all the setup and teardown steps manually and updated CI as well

lorenyu commented:
not entirely sure why CI fails. it seems to work when i run all the steps manually. any help would be appreciated. it looks like the error in CI is

> Waiter ServicesStable failed: Waiter encountered a terminal failure state: For expression "services[].status" we matched expected path: "INACTIVE" at least once

Nava-JoshLong commented:
Haven't had a chance to dive too deep into this, but I did add a feature in the NJ Makefile I thought could be useful in other spots here. You can wildcard a part of a command to specify the environment, `make` checks if that is an allowed environment, and then runs. Here's a quick example

```
ALLOWED_ENV_LIST := dev test prod # Allowed environments to deploy to

check-env: ## Checks if the environment passed is an allowed environment to deploy to
ifeq (,$(findstring $(env),$(ALLOWED_ENV_LIST)))
	$(error $(env) is not supported)
else
	@echo "$(env) is a supported environment"
endif

app-init-%: ## Runs terraform init in the AWS account specified
	$(MAKE) check-env env=$*
	@echo "Deploying the application infra to $*"
	cd deploy/application && $(MAKE) init-$*
```

So, `make app-init-dev` would work, but `make app-init-dev2` wouldn't since it isn't in the allowed list. In the directory that it changes to, I have the backend setup via runtime flag instead of a file like in this PR. You could add the table and region like this as well. I'm interested to hear thoughts on doing it like this vs using files like in this PR

```
init-%: ## Runs init on the environment specified
	terraform init -var="environment=$*" \
		-backend-config="bucket=bucket-name-$*-tf-state-us-east-1"

plan-%: ## Runs plan on the environment specified
	$(MAKE) init-$*
	terraform plan -var="environment=$*"

apply-%: ## Runs apply on the environment specified
	$(MAKE) init-$*
	terraform apply -var="environment=$*"
```

lorenyu commented:
Thanks @Nava-JoshLong!

I thought of using make target `%` wildcard to represent the environment but decided against it since make targets can only have one wildcard and some targets have more than one parameter – in particular, for multi-app repos we want `APP_NAME` and `ENVIRONMENT` to both be parameters, so instead of `make infra-app-service-dev` it's `make infra-app-service APP_NAME=app ENVIRONMENT=dev`.

I like the files because backends are so fundamental that I'd like to keep them static and not do any fancy dynamic generation of the backend configs. I also want them in source control and for any changes to show up explicitly in PRs.

lorenyu commented on infra/modules/terraform-backend-s3/main.tf:
```diff
@@ -269,10 +274,3 @@ resource "aws_s3_bucket_logging" "tf_state" {
   target_bucket = aws_s3_bucket.tf_log.id
   target_prefix = "logs/${aws_s3_bucket.tf_state.bucket}/"
 }
-
-resource "aws_s3_bucket_logging" "tf_log" {
-  bucket = aws_s3_bucket.tf_log.id
-
-  target_bucket = aws_s3_bucket.tf_log.id
-  target_prefix = "logs/${aws_s3_bucket.tf_log.bucket}/"
-}
```
removed this for now because i'm running into a super cryptic error that only happens when running in GitHub actions (doesn't happen when i run CI locally)

![image](https://user-images.githubusercontent.com/447859/230695628-e707dc0e-95f6-447c-955e-456a8e714c6b.png)

see github run that failed: https://github.com/navapbc/template-infra/actions/runs/4641980887/jobs/8215462775

shawnvanderjagt commented on docs/decisions/infra/separate-terraform-backend-configs-into-separate-config-files.md:
```diff
@@ -0,0 +1,12 @@
+For creating accounts, can't use the .tfbackend backend config file approach because the main.tf file can only have one backend configuration, so if we have the backend configuration as a partial configuration of `backend "s3" {}`, then we can't use that same module to configure a new account, since the process for configuring a new account
+requires setting the backend configuration to `backend "local" {}`. We could have a separate duplicate module that's has backend set to local. or we could also temporarily update the backend from `"s3"` to `"local"`, but both of those approaches seem confusing.
+
+Another alternative is to go back to the old way of bootstrapping an account i.e. to do it via a script that creates an S3 bucket via AWS CLI. The bootstrap script would only do the minimal configuration for the S3 bucket, and let terraform handle the remainder of the configuration, such as creating the dynamodb tables. At this point, there is no risk of not having state locking in place since th ...
```
We could use the CLI method coupled with [Import](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/s3_bucket#import)
to get that initial bucket back under management by terraform. 
[Guide](https://terrateam.io/blog/aws-s3-import)

I haven't fully fleshed this thought out. But it is something we could consider in the future.

shawnvanderjagt commented on bin/set-up-current-account.sh:
```diff
@@ -0,0 +1,63 @@
+ACCOUNT_ALIAS="$(./bin/current-account-alias.sh)"
+ACCOUNT_ID="$(./bin/current-account-id.sh)"
+REGION="$(./bin/current-region.sh)"
+BACKEND_CONFIG_FILE="$ACCOUNT_ALIAS.s3.tfbackend"
```
I don't think `BACKEND_CONFIG_FILE` variable is being used. Seems like its being handled on line 23 of `bin/create-tfbackend.sh` which uses the `BACKEND_CONFIG_NAME` passed as an argument on line 62 of this file.

shawnvanderjagt commented on docs/infra/set-up-app-build-repository.md:
````diff
@@ -0,0 +1,34 @@
+Now run the following commands to create the resources, making sure to verify the plan before confirming the apply.
+
+```bash
+make infra-build-repository APP_NAME=app
````
I think this should be make infra-app-build-repository APP_NAME=app

shawnvanderjagt replied:
Or possibly the command in the Makefile on line 48 needs to be updated

shawnvanderjagt commented on Makefile:
```diff
@@ -31,6 +33,25 @@ ENVIRONMENTS := $(notdir $(wildcard infra/app/envs/*))
+infra-account:
+	./bin/terraform-init-and-apply.sh infra/accounts $(ACCOUNT_ALIAS) $(TF_APPLY_ARGS)
+
+infra-app-build-repository:
```
See note in in set-up-app-build-repossitory.md

shawnvanderjagt commented on Makefile:
```diff
@@ -100,10 +121,10 @@ release-publish:
 
 release-deploy:
 # check the varaible against the list of enviroments and suggest one of the correct envs.
```
variable is misspelled here.

shawnvanderjagt commented on docs/decisions/infra/separate-terraform-backend-configs-into-separate-config-files.md:
```diff
@@ -0,0 +1,12 @@
+# Notes
+
+For creating accounts, can't use the .tfbackend backend config file approach because the main.tf file can only have one backend configuration, so if we have the backend configuration as a partial configuration of `backend "s3" {}`, then we can't use that same module to configure a new account, since the process for configuring a new account
```
For creating accounts, we can't use

lorenyu replied:
@shawnvanderjagt Sorry, the previous version of this markdown doc was just some notes that I had, not a properly written ADR that was ready for review. Anyways, I rewrote the ADR and now it's completely done. Take another look

shawnvanderjagt commented on docs/decisions/infra/separate-terraform-backend-configs-into-separate-config-files.md:
```diff
@@ -0,0 +1,12 @@
+# Notes
+
+For creating accounts, can't use the .tfbackend backend config file approach because the main.tf file can only have one backend configuration, so if we have the backend configuration as a partial configuration of `backend "s3" {}`, then we can't use that same module to configure a new account, since the process for configuring a new account
+requires setting the backend configuration to `backend "local" {}`. We could have a separate duplicate module that's has backend set to local. or we could also temporarily update the backend from `"s3"` to `"local"`, but both of those approaches seem confusing.
```
module that has backend set to local or we could temporarily

kyeah reviewed (commented):
I didn't get through everything, but adding my initial comments below 🙂 thanks for dedicating time for this refactor!

shawnvanderjagt commented on bin/set-up-current-account.sh:
```diff
@@ -0,0 +1,63 @@
+set -euo pipefail
+
+# Name the account based on the current account alias
+ACCOUNT_ALIAS="$(./bin/current-account-alias.sh)"
```
Something to consider, I acquired a fresh account that did not have an alias setup initially, of course this broke this script until I setup that alias. It may be good to add a warning that an alias is needed or if an alias is not found to use the account id instead.

lorenyu replied:
Thanks! I was actually wondering if all accounts have an alias.

What do you think we should do. I like the idea of being robust and just using account id, but from a developer experience perspective it's really nice to be able to see a folder with names like `project-dev.tfbackend, project-prod.tfbackend` rather than `123415345.tfbackend`, `23452362.tfbackend`.

If we fallback to id, then after the account is configured, if an alias is added, they'd need to rename the file to the alias otherwise it won't work.

This makes me wonder:
1. Can an account have two or more aliases?
2. Can an account change their alias?

What are your thoughts?

shawnvanderjagt replied:
1. An account can have only a single alias
2. As the root user of my account I was able to set the alias. I also found out recently a fresh bid account was actually created for mcads but fell through the cracks somehow. In that account I updated the alias from `bid-fed-mcads-tech-challenge` to `bid-tech-challenge` and transferred that account to Haiyan, so it is possible. There is also a miss named account alias in my projects greenfield project that will need to be updated. 

I agree that `prod.tfbackend` is easier on the brain than `123456789012.tfbackend`. I am just not sure how to make it easy to read and guarantee its correct.  I have been going back and forth between the following:
1. Requiring template users to ensure they have an account alias and provide instructions for if it changes.
2. Just using the account_id since once setup is complete there wont be a lot of changes happening. 
3. Allowing the user to name their own backend file and refactoring for that. 
4. Combine 2 & 3 using some kind of “symlink” to link them or including `myname.account_id.s3.tfbackend`.

lorenyu replied:
I have an idea, for accounts, how about when we create the backend config file in `set-up-current-account.sh`, we name the file `<alias>.<account-id>.tfbackend`, where `<alias>` is a user specified alias. It can be the account alias, or whatever, but it's not actually used. Perhaps it can default to the account alias if it exists. When running things, in order to find the right tfbackend file, we can do something like `ls | grep <account-id>`.

Since running things only uses the account-id, the alias is purely for readability when looking at the file system, so someone can rename the file if they want to whatever is most readable.

lorenyu replied:
I went ahead and implemented that last idea. Take a look and lemme know what you think!

shawnvanderjagt replied:
That's good I like it.

shawnvanderjagt commented on bin/set-up-current-account.sh:
```diff
@@ -0,0 +1,63 @@
+  -backend-config="region=$REGION"
+
+# Import the S3 bucket that was created in the previous step so we don't recreate it
+terraform import module.backend.aws_s3_bucket.tf_state $TF_STATE_BUCKET_NAME
```
If the bucket already exists when this script is run you will take an error on import.

lorenyu replied:
What do you suggest? Should we do a check to see if the bucket exists at the beginning of the script?

shawnvanderjagt replied:
It technically didn't break anything so something as simple as "this is expected if you already created this resource" above the error would be sufficient. If not we would want to check that both the bucket exists and the backend file exists and matches the bucket  before trying to direct them to the best course of action.

lorenyu replied:
Hmm, I'm not following. How do I reproduce the issue you're describing? Are you referring to the situation where someone runs `set-up-current-account.sh` twice, or the situation where by some coincidence there's already a bucket with the same name that some other person in the world created, and now the name conflicts?

shawnvanderjagt replied:
You can reproduce the issue by running the `set-up-current-account.sh` script more than once. The most likely scenario probably be the same person running it twice.

lorenyu replied:
I see, I get this:

<img width="750" alt="image" src="https://github.com/navapbc/template-infra/assets/447859/78060a95-2df0-451b-83e5-efd1d8d29264">

seems like a reasonable error for now, but I'll create a follow-up ticket to be a little more defensive in case someone runs into this.

lorenyu replied:
https://github.com/navapbc/template-infra/issues/268

lorenyu commented:
@shawnvanderjagt @kyeah thanks for the initial reviews! i'll work on addressing the feedback as i get caught up on things.

in the meantime, what do you both think about the overall approach as the default approach for future infra? kevin, i remembered from the slack discussion that you wanted to see an example of how things look before you committed to whether you like the direction.

i wanted to be sure since it'll be a little hard to go back if we go forward with this direction, and it'll impact development of future work (in particular the database stuff i'm working on), so it'd be good to have a firm direction.

lorenyu commented on bin/terraform-apply.sh:
```diff
@@ -0,0 +1,42 @@
+# -----------------------------------------------------------------------------
+# Convenience script for running terraform apply for the specified module and configuration name.
+# The configuration name is used to determine which .tfvars file to use for the -var-file
+# option of terraform apply. Additional arguments to terraform apply can also be passed in.
```
Yeah that sounds nice. Only thing is that autocomplete and variables is a bit annoying with Makefiles, where it's like `make target ARGS="some string of args"` rather than `make target some args`, which is why I gave the common make commands specific make targets so that ohmyzsh can autocomplete when you type `make infra-<tab>`.

What do you think about doing that in a follow up ticket?

And what do you think about the instructions for running terraform directly in https://github.com/navapbc/template-infra/pull/242/files#diff-546e28d202b614e4cbc4c27f03c9bda9a502d143e5cdb167a87d47feb545c67dR51 ? Any ideas for improving those?

lorenyu commented on Makefile:
```diff
@@ -31,6 +33,25 @@ ENVIRONMENTS := $(notdir $(wildcard infra/app/envs/*))
+infra-configure-app-service:
+	./bin/configure-app-service.sh $(APP_NAME) $(ENVIRONMENT)
+
+infra-account:
```
sure that makes sense to me. i originally was thinking that `infra-init-and-apply-account` is too long, but `update` i think is short enough and meets the need as effectively i think.

lorenyu commented on Makefile:
```diff
@@ -6,6 +6,8 @@ PROJECT_NAME ?= $(notdir $(PWD))
 # on will be determined by the APP_NAME Makefile argument
 APP_NAME ?= app
 
+ACCOUNT_ALIAS := `./bin/current-account-alias.sh`
```
Hmm, that's a good point. I don't like that. I swapped it with `=` so that the scripted doesn't get executed unless the variable is used/expanded.

lorenyu commented on template-only-bin/set-up-project.sh:
```diff
@@ -34,3 +34,6 @@ sed -i.bak "s|<REPO_URL>|$REPO_URL|" main.tf
 sed -i.bak "s/<DEFAULT_REGION>/$DEFAULT_REGION/" main.tf
+
+# Remove the backup file created by sed
+rm main.tf.bak
```
It's because `sed` behaves differently on Mac and github actions due to different versions of sed being used. On one of the versions, the `-i` flag is required in order to edit files in place. Without `-i` the files always end up blank.

Should I document that as a comment? Not sure where to put that comment. Does it make sense to put it here, or up above where we pass in the `-i` flag?

lorenyu commented on bin/terraform-apply.sh:
```diff
@@ -0,0 +1,42 @@
+#   For application modules the config name is the name of the environment (e.g. "dev", "staging", "prod").
+#   For application modules that are shared across environments, the config name is "shared".
+#   For example if a backend config file is named "myaccount.s3.tfbackend", then the CONFIG_NAME would be "myaccount"
+# TF_APPLY_ARGS (optional) – Any additional arguments to pass to terraform apply.
```
Oh yeah good point thanks. I'll change it.

lorenyu commented on bin/terraform-init.sh:
```diff
@@ -0,0 +1,30 @@
+
+# 1. Set working directory to the terraform root module directory
+
+cd $MODULE_DIR
```
yeah. for some reason i find it confusing to use `-chdir` and `-backend-config` options together since it's a little ambiguous to me whether the config file is a path relative to the directory the command was run in or the directory terraform changed to. this way was more obvious to me.

but now that i think about it, i'll just use -chdir but add a comment explaining that the config is relative to the new directory not the current one
